### PR TITLE
feat: add welcome signup email

### LINKED ISSUE #2971: email invite formatting issues
- [ ] Email template is managed in code so we can audit + approve changes (consider plaintext for now for easy rendering)
- [ ] Off-centered text: Screen Shot 2022-08-18 at 3 08 06 PM
- [ ] Remove "unsubscribe" since it's a transactional email

### PR DESCRIPTION
**Pull request by ssoroka**
## Summary

- switches to SMTP for email delivery instead of sendgrid api
- renders email templates locally instead of depending on remote service
- switches existing emails over to new templates
- adds welcome/signup email

## Checklist

- [x] Wrote appropriate unit tests
- [x] Considered security implications of the change
- [x] Updated associated docs where necessary
- [x] Updated associated configuration where necessary
- [x] Change is backwards compatible if it needs to be (user can upgrade without manual steps?)
- [x] Nothing sensitive logged
- [x] Considered data migrations for smooth upgrades

also..
Resolves #2971

### CODE REVIEW

File: .golangci.yml [hunk 1]
```diff
@@ -47,10 +46,8 @@ linters:
     - rowserrcheck
     - sqlclosecheck
     - staticcheck
-    - structcheck
     - typecheck
     - unused
-    - varcheck
```
Comment (ssoroka) [inline .golangci.yml:53]: these are deprecated in favor of unused, which we already use.

File: internal/generate/generate.go [hunk 2]
```diff
@@ -25,6 +25,8 @@ func CryptoRandom(n int, charset string) (string, error) {
 
 	bytes := make([]byte, n)
 	for i := range bytes {
+		// linter is mistaken about which package this is
```
Comment (ssoroka) [inline internal/generate/generate.go:28]: seems like a 1.49 linter bug

File: .golangci.yml (hunk 1 above)
Comment (ssoroka) [inline .golangci.yml:53]: these are deprecated in favor of `unused`, which we already use

File: internal/server/email/smtp.go [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+	}
+	c.Conn = conn
+
+	defer func() {
+		if err != nil {
+			c.Conn = nil
+		}
+	}()
+
+	c.reader = bufio.NewReader(c.Conn)
```
Comment (pdevine) [inline internal/server/email/smtp.go:63]: Why not use use "net/smtp" here? I'm worried this code might be fragile.

I think you can do something like:
```
    auth := smtp.PlainAuth("","mail@infrahq.com", "password", host)
    
    client, err := smtp.NewClient(conn, host)
    if err != nil {
        log.Panic(err)
    }

    // Auth
    if err = client.Auth(auth); err != nil {
        log.Panic(err)
    }
```

File: internal/server/email/smtp.go [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+		}
+	}
+	if len(msg.Subject) > 0 {
+		if err := client.writeln(fmt.Sprintf("Subject: %s", msg.Subject)); err != nil {
+			return fmt.Errorf("write subject: %w", err)
+		}
+	}
+
+	// mime multipart
+	if err := client.writeln("MIME-Version: 1.0\r\nContent-Type: multipart/alternative; boundary=c3VwYWhpbmZyYQ\r\n"); err != nil {
```
Comment (BruceMacD) [inline internal/server/email/smtp.go:197]: What does the `boundary` do?

Comment (ssoroka) [reply internal/server/email/smtp.go:197]: part of how MIME formatting works, it tells the smtp server where each section starts and stops

File: internal/server/email/smtp.go [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+// SendSMTP sends an email message
+func SendSMTP(msg Message) error {
+	client.Lock()
+	defer client.Unlock()
+
+	if err := client.connect(); err != nil {
+		return err
+	}
+
+	if err := client.writeln(fmt.Sprintf("MAIL FROM: %s", msg.FromAddress)); err != nil {
```
Comment (pdevine) [inline internal/server/email/smtp.go:142]: Why not use `smtp.SendMail()` instead?

Comment (ssoroka) [reply internal/server/email/smtp.go:142]: it was just so simple. :D I can change it though.

Comment (ssoroka) [reply internal/server/email/smtp.go:142]: So one problem I'm running into with smtp.SendMail is that it assumes :smtp port (25) and then tries to upgrade to tls if the server supports it. This is different from what I'm doing; I'm connecting to a TLS-encrypted SMTP port and failing if it doesn't connect. Switching to this other code means there's a potential silent downgrade to non-TLS if the destination server doesn't support TLS

**Review by pdevine — APPROVED** [review]
LGTM

**Pull request merged**